src/grad/variable.cr

```diff
@@ -205,6 +205,39 @@ class Num::Grad::Variable(T)
     result
   end
 
+  # Computes the sum of all the elements of the variable
+  #
+  # ## Examples
+  #
+  # ```
+  # ctx = Num::Grad::Context(Tensor(Float64, CPU(Float64))).new
+  # x = ctx.variable([1.0, 2.0, 3.0, 4.0])
+  # x.sum # => [10.0]
+  # ```
+  def sum : Num::Grad::Variable(T)
+    result = @context.variable(Num.sum(@value, axis: -1, dims: true))
```

Comment:
This method signature isn't quite accurate for what is being done here.  This is reducing along the final axis rather than reducing to a scalar, which is going to cause unexpected behavior in your `mean` method for higher dimensions since it will be using the size of the entire `Tensor` rather than per axis.  If the goal is forward/backward for a total summation, `@value.sum` should be what you need (which is not currently supported by the OCL backend).

Reply:
Good catch. I was OK with the `#sum` reducing along the final axis; but I didn't think through the impact on `#size` when implementing `#mean`. Ok, let me see what I can do.

Reply:
@christopherzimmerman, I've modified the `#sum` and `#mean` methods to accept `axis` as a parameter. And I've refactored the tests and expanded them to test for 1D and 2D tensor variables.